### PR TITLE
Features/rollback tss

### PR DESCRIPTION
**Pull request by byteflyfunny**
add new rollback logic when mpc validators verify failed

### CODE REVIEW

File: l2geth/core/vm/evm.go [hunk 1]
```diff
@@ -195,6 +196,11 @@ func (evm *EVM) Interpreter() Interpreter {
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	//TODO test roll back
+	if addr == dump.BvmReorgAddress {
```
Comment (silmaj) [inline l2geth/core/vm/evm.go:200]: i guess we don't need this?

File: l2geth/rollup/dump/constants.go [hunk 2]
```diff
@@ -9,3 +9,4 @@ var BvmFeeWallet = common.HexToAddress("0x42000000000000000000000000000000000000
 var TssRewardAddress = common.HexToAddress("0x4200000000000000000000000000000000000020")
 var DeadAddress = common.HexToAddress("0xdeaDDeADDEaDdeaDdEAddEADDEAdDeadDEADDEaD")
 var BvmWhitelistAddress = common.HexToAddress("0x4200000000000000000000000000000000000002")
+var BvmReorgAddress = common.HexToAddress("0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD2222")
```
Comment (silmaj) [inline l2geth/rollup/dump/constants.go:12]: same above

File: packages/contracts/contracts/L1/rollup/IStateCommitmentChain.sol [hunk 3]
```diff
@@ -86,4 +88,19 @@ interface IStateCommitmentChain {
...
     view
     returns (bool _inside);
+
+    /**
+     * Emit event to notify sequencers to roll back.
+     * @param _shouldRollBack roll back to should start .
+     * @param _shouldStartAtElement Index of the element at which this batch should start
+     * @param _signature signature of rollback message
+     */
+    function rollBackL2Chain(uint256 _shouldRollBack,uint256 _shouldStartAtElement, bytes memory _signature) external;
```
Comment (silmaj) [inline packages/contracts/contracts/L1/rollup/IStateCommitmentChain.sol:98]: do we need a signature for rollback?

File: packages/contracts/contracts/L1/rollup/StateCommitmentChain.sol [hunk 4]
```diff
@@ -181,6 +181,36 @@ contract StateCommitmentChain is IStateCommitmentChain, Lib_AddressResolver, Cro
...
+        _checkRollBackSignature(_shouldRollBack,_signature);
+
+    }
+
+
+    /**
+    * @inheritdoc IStateCommitmentChain
+     */
+    // slither-disable-next-line external-function
+    function rollBackMessage(uint256 _shouldRollBack) public {}
```
Comment (silmaj) [inline packages/contracts/contracts/L1/rollup/StateCommitmentChain.sol:211]: elimit this

File: l2geth/core/vm/evm.go (hunk 1 above)
Comment (byteflyfunny) [reply l2geth/core/vm/evm.go:200]: yes, TODO first

File: packages/contracts/contracts/L1/rollup/StateCommitmentChain.sol (hunk 4 above)
Comment (byteflyfunny) [reply packages/contracts/contracts/L1/rollup/StateCommitmentChain.sol:211]: yep, we emit event like this

**Review by guoshijiang — APPROVED** [review]

**Review by curryxbo — APPROVED** [review]

**Review by silmaj — APPROVED** [review]

**Pull request merged**